File: pkg/templates/python/yutori-computer-use/tools/computer.py
"""
Yutori n1 Computer Tool

Maps n1-latest action format to Kernel's Computer Controls API.
Screenshots are converted to WebP for better compression across multi-step trajectories.
"""

import asyncio
import base64
from io import BytesIO
from typing import Literal, TypedDict

from kernel import Kernel
from PIL import Image

from .base import ToolError, ToolResult

TYPING_DELAY_MS = 12
SCREENSHOT_DELAY_S = 0.3
ACTION_DELAY_S = 0.3

N1ActionType = Literal[
    "left_click",
    "double_click",
    "triple_click",
    "right_click",
    "scroll",
    "type",
    "key_press",
    "hover",
    "drag",
    "wait",
    "refresh",
    "go_back",
    "goto_url",
]


class N1Action(TypedDict, total=False):
    action_type: N1ActionType
    coordinates: tuple[int, int] | list[int]
    start_coordinates: tuple[int, int] | list[int]
    direction: Literal["up", "down", "left", "right"]
    amount: int
    text: str
    press_enter_after: bool
    clear_before_typing: bool
    key_comb: str
    url: str


KEY_MAP = {
    "Enter": "Return",
    "Escape": "Escape",
    "Backspace": "BackSpace",
    "Tab": "Tab",
    "Delete": "Delete",
    "ArrowUp": "Up",
    "ArrowDown": "Down",
    "ArrowLeft": "Left",
    "ArrowRight": "Right",
    "Home": "Home",
    "End": "End",
    "PageUp": "Page_Up",
    "PageDown": "Page_Down",
    "F1": "F1",
    "F2": "F2",
    "F3": "F3",
    "F4": "F4",
    "F5": "F5",
    "F6": "F6",
    "F7": "F7",
    "F8": "F8",
    "F9": "F9",
    "F10": "F10",
    "F11": "F11",
    "F12": "F12",
}

MODIFIER_MAP = {
    "control": "ctrl",
    "ctrl": "ctrl",
    "alt": "alt",
    "shift": "shift",
    "meta": "super",
    "command": "super",
    "cmd": "super",
}


class ComputerTool:
    def __init__(self, kernel: Kernel, session_id: str, width: int = 1280, height: int = 800):
        self.kernel = kernel
        self.session_id = session_id
        self.width = width
        self.height = height

    async def execute(self, action: N1Action) -> ToolResult:
        action_type = action.get("action_type")

        handlers = {
            "left_click": lambda a: self._handle_click(a, "left", 1),
            "double_click": lambda a: self._handle_click(a, "left", 2),
            "triple_click": lambda a: self._handle_click(a, "left", 3),
            "right_click": lambda a: self._handle_click(a, "right", 1),
            "scroll": self._handle_scroll,
            "type": self._handle_type,
            "key_press": self._handle_key_press,
            "hover": self._handle_hover,
            "drag": self._handle_drag,
            "wait": self._handle_wait,
            "refresh": self._handle_refresh,
            "go_back": self._handle_go_back,
            "goto_url": self._handle_goto_url,
        }

        handler = handlers.get(action_type)
        if not handler:
            raise ToolError(f"Unknown action type: {action_type}")

        return await handler(action)

    async def _handle_click(self, action: N1Action, button: str, num_clicks: int) -> ToolResult:
        coords = self._get_coordinates(action.get("coordinates"))

        self.kernel.browsers.computer.click_mouse(
            self.session_id,
            x=coords["x"],
            y=coords["y"],
            button=button,
            click_type="click",
            num_clicks=num_clicks,
        )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_scroll(self, action: N1Action) -> ToolResult:
        coords = self._get_coordinates(action.get("coordinates"))
        direction = action.get("direction")
        amount = action.get("amount", 3)

        if direction not in ("up", "down", "left", "right"):
            raise ToolError(f"Invalid scroll direction: {direction}")

        scroll_delta = amount * 100

        delta_x = 0
        delta_y = 0

        if direction == "up":
            delta_y = -scroll_delta
        elif direction == "down":
            delta_y = scroll_delta
        elif direction == "left":
            delta_x = -scroll_delta
        elif direction == "right":
            delta_x = scroll_delta

        self.kernel.browsers.computer.scroll(
            self.session_id,
            x=coords["x"],
            y=coords["y"],
            delta_x=delta_x,
            delta_y=delta_y,
        )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_type(self, action: N1Action) -> ToolResult:
        text = action.get("text")
        if not text:
            raise ToolError("text is required for type action")

        if action.get("clear_before_typing"):
            self.kernel.browsers.computer.press_key(
                self.session_id,
                keys=["ctrl+a"],
            )
            await asyncio.sleep(0.1)
            self.kernel.browsers.computer.press_key(
                self.session_id,
                keys=["BackSpace"],
            )
            await asyncio.sleep(0.1)

        self.kernel.browsers.computer.type_text(
            self.session_id,
            text=text,
            delay=TYPING_DELAY_MS,
        )

        if action.get("press_enter_after"):
            await asyncio.sleep(0.1)
            self.kernel.browsers.computer.press_key(
                self.session_id,
                keys=["Return"],
            )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_key_press(self, action: N1Action) -> ToolResult:
        key_comb = action.get("key_comb")
        if not key_comb:
            raise ToolError("key_comb is required for key_press action")

        mapped_key = self._map_key(key_comb)

        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=[mapped_key],
        )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_hover(self, action: N1Action) -> ToolResult:
        coords = self._get_coordinates(action.get("coordinates"))

        self.kernel.browsers.computer.move_mouse(
            self.session_id,
            x=coords["x"],
            y=coords["y"],
        )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_drag(self, action: N1Action) -> ToolResult:
        start_coords = self._get_coordinates(action.get("start_coordinates"))
        end_coords = self._get_coordinates(action.get("coordinates"))

        self.kernel.browsers.computer.drag_mouse(
            self.session_id,
            path=[[start_coords["x"], start_coords["y"]], [end_coords["x"], end_coords["y"]]],
            button="left",
        )

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_wait(self, action: N1Action) -> ToolResult:
        await asyncio.sleep(2)
        return await self.screenshot()

    async def _handle_refresh(self, action: N1Action) -> ToolResult:
        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=["F5"],
        )
        await asyncio.sleep(2)
        return await self.screenshot()

    async def _handle_go_back(self, action: N1Action) -> ToolResult:
        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=["alt+Left"],
        )
        await asyncio.sleep(1.5)
        return await self.screenshot()

    async def _handle_goto_url(self, action: N1Action) -> ToolResult:
        url = action.get("url")
        if not url:
            raise ToolError("url is required for goto_url action")

        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=["ctrl+l"],
        )
        await asyncio.sleep(ACTION_DELAY_S)

        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=["ctrl+a"],
        )
        await asyncio.sleep(0.1)

        self.kernel.browsers.computer.type_text(
            self.session_id,
            text=url,
            delay=TYPING_DELAY_MS,
        )
        await asyncio.sleep(ACTION_DELAY_S)

        self.kernel.browsers.computer.press_key(
            self.session_id,
            keys=["Return"],
        )
        await asyncio.sleep(2)
        return await self.screenshot()

    async def screenshot(self) -> ToolResult:
        try:
            response = self.kernel.browsers.computer.capture_screenshot(
                self.session_id
            )
            png_bytes = response.read()
            img = Image.open(BytesIO(png_bytes))
            webp_buf = BytesIO()
            img.save(webp_buf, "WEBP", quality=80)
            base64_image = base64.b64encode(webp_buf.getvalue()).decode("utf-8")
            return {"base64_image": base64_image}
        except Exception as e:
            raise ToolError(f"Failed to take screenshot: {e}")

    def _get_coordinates(
        self, coords: tuple[int, int] | list[int] | None
    ) -> dict[str, int]:
        if coords is None or len(coords) != 2:
            return {"x": self.width // 2, "y": self.height // 2}

        x, y = coords
        if not isinstance(x, (int, float)) or not isinstance(y, (int, float)) or x < 0 or y < 0:
            raise ToolError(f"Invalid coordinates: {coords}")

        return {"x": int(x), "y": int(y)}

    def _map_key(self, key: str) -> str:
        if "+" in key:
            parts = key.split("+")
            mapped_parts = []
            for part in parts:
                trimmed = part.strip()
                lower = trimmed.lower()
                
                if lower in MODIFIER_MAP:
                    mapped_parts.append(MODIFIER_MAP[lower])
                else:
                    mapped_parts.append(KEY_MAP.get(trimmed, trimmed))
            
            return "+".join(mapped_parts)

        return KEY_MAP.get(key, key)

File: pkg/templates/typescript/yutori-computer-use/tools/computer.ts
/**
 * Yutori n1 Computer Tool
 * 
 * Maps n1-latest action format to Kernel's Computer Controls API.
 * Screenshots are converted to WebP for better compression across multi-step trajectories.
 */

import { Buffer } from 'buffer';
import type { Kernel } from '@onkernel/sdk';
import sharp from 'sharp';

const TYPING_DELAY_MS = 12;
const SCREENSHOT_DELAY_MS = 300;
const ACTION_DELAY_MS = 300;

export interface ToolResult {
  base64Image?: string;
  output?: string;
  error?: string;
}

export class ToolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ToolError';
  }
}

export type N1ActionType =
  | 'left_click'
  | 'double_click'
  | 'triple_click'
  | 'right_click'
  | 'scroll'
  | 'type'
  | 'key_press'
  | 'hover'
  | 'drag'
  | 'wait'
  | 'refresh'
  | 'go_back'
  | 'goto_url';

export interface N1Action {
  action_type: N1ActionType;
  coordinates?: [number, number];
  start_coordinates?: [number, number];
  direction?: 'up' | 'down' | 'left' | 'right';
  amount?: number;
  text?: string;
  press_enter_after?: boolean;
  clear_before_typing?: boolean;
  key_comb?: string;
  url?: string;
}

const KEY_MAP: Record<string, string> = {
  'Enter': 'Return',
  'Escape': 'Escape',
  'Backspace': 'BackSpace',
  'Tab': 'Tab',
  'Delete': 'Delete',
  'ArrowUp': 'Up',
  'ArrowDown': 'Down',
  'ArrowLeft': 'Left',
  'ArrowRight': 'Right',
  'Home': 'Home',
  'End': 'End',
  'PageUp': 'Page_Up',
  'PageDown': 'Page_Down',
  'F1': 'F1',
  'F2': 'F2',
  'F3': 'F3',
  'F4': 'F4',
  'F5': 'F5',
  'F6': 'F6',
  'F7': 'F7',
  'F8': 'F8',
  'F9': 'F9',
  'F10': 'F10',
  'F11': 'F11',
  'F12': 'F12',
};

const MODIFIER_MAP: Record<string, string> = {
  'control': 'ctrl',
  'ctrl': 'ctrl',
  'alt': 'alt',
  'shift': 'shift',
  'meta': 'super',
  'command': 'super',
  'cmd': 'super',
};

export class ComputerTool {
  private kernel: Kernel;
  private sessionId: string;
  private width: number;
  private height: number;
  private kioskMode: boolean;

  constructor(kernel: Kernel, sessionId: string, width = 1280, height = 800, kioskMode = false) {
    this.kernel = kernel;
    this.sessionId = sessionId;
    this.width = width;
    this.height = height;
    this.kioskMode = kioskMode;
  }

  async execute(action: N1Action): Promise<ToolResult> {
    const { action_type } = action;

    switch (action_type) {
      case 'left_click':
        return this.handleClick(action, 'left', 1);
      case 'double_click':
        return this.handleClick(action, 'left', 2);
      case 'triple_click':
        return this.handleClick(action, 'left', 3);
      case 'right_click':
        return this.handleClick(action, 'right', 1);
      case 'scroll':
        return this.handleScroll(action);
      case 'type':
        return this.handleType(action);
      case 'key_press':
        return this.handleKeyPress(action);
      case 'hover':
        return this.handleHover(action);
      case 'drag':
        return this.handleDrag(action);
      case 'wait':
        return this.handleWait();
      case 'refresh':
        return this.handleRefresh();
      case 'go_back':
        return this.handleGoBack();
      case 'goto_url':
        return this.handleGotoUrl(action);
      default:
        throw new ToolError(`Unknown action type: ${action_type}`);
    }
  }

  private async handleClick(action: N1Action, button: 'left' | 'right', numClicks: number): Promise<ToolResult> {
    const coords = this.getCoordinates(action.coordinates);
    
    await this.kernel.browsers.computer.clickMouse(this.sessionId, {
      x: coords.x,
      y: coords.y,
      button,
      click_type: 'click',
      num_clicks: numClicks,
    });

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleScroll(action: N1Action): Promise<ToolResult> {
    const coords = this.getCoordinates(action.coordinates);
    const direction = action.direction;
    const amount = action.amount ?? 3;

    if (!direction || !['up', 'down', 'left', 'right'].includes(direction)) {
      throw new ToolError(`Invalid scroll direction: ${direction}`);
    }

    const scrollDelta = amount * 100;

    let delta_x = 0;
    let delta_y = 0;

    switch (direction) {
      case 'up':
        delta_y = -scrollDelta;
        break;
      case 'down':
        delta_y = scrollDelta;
        break;
      case 'left':
        delta_x = -scrollDelta;
        break;
      case 'right':
        delta_x = scrollDelta;
        break;
    }

    await this.kernel.browsers.computer.scroll(this.sessionId, {
      x: coords.x,
      y: coords.y,
      delta_x,
      delta_y,
    });

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleType(action: N1Action): Promise<ToolResult> {
    const text = action.text;
    if (!text) {
      throw new ToolError('text is required for type action');
    }

    if (action.clear_before_typing) {
      await this.kernel.browsers.computer.pressKey(this.sessionId, {
        keys: ['ctrl+a'],
      });
      await this.sleep(100);
      await this.kernel.browsers.computer.pressKey(this.sessionId, {
        keys: ['BackSpace'],
      });
      await this.sleep(100);
    }

    await this.kernel.browsers.computer.typeText(this.sessionId, {
      text,
      delay: TYPING_DELAY_MS,
    });

    if (action.press_enter_after) {
      await this.sleep(100);
      await this.kernel.browsers.computer.pressKey(this.sessionId, {
        keys: ['Return'],
      });
    }

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleKeyPress(action: N1Action): Promise<ToolResult> {
    const keyComb = action.key_comb;
    if (!keyComb) {
      throw new ToolError('key_comb is required for key_press action');
    }

    const mappedKey = this.mapKey(keyComb);

    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: [mappedKey],
    });

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleHover(action: N1Action): Promise<ToolResult> {
    const coords = this.getCoordinates(action.coordinates);

    await this.kernel.browsers.computer.moveMouse(this.sessionId, {
      x: coords.x,
      y: coords.y,
    });

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleDrag(action: N1Action): Promise<ToolResult> {
    const startCoords = this.getCoordinates(action.start_coordinates);
    const endCoords = this.getCoordinates(action.coordinates);

    await this.kernel.browsers.computer.dragMouse(this.sessionId, {
      path: [[startCoords.x, startCoords.y], [endCoords.x, endCoords.y]],
      button: 'left',
    });

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleWait(): Promise<ToolResult> {
    await this.sleep(2000);
    return this.screenshot();
  }

  private async handleRefresh(): Promise<ToolResult> {
    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: ['F5'],
    });

    await this.sleep(2000);
    return this.screenshot();
  }

  private async handleGoBack(): Promise<ToolResult> {
    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: ['alt+Left'],
    });

    await this.sleep(1500);
    return this.screenshot();
  }

  private async handleGotoUrl(action: N1Action): Promise<ToolResult> {
    const url = action.url;
    if (!url) {
      throw new ToolError('url is required for goto_url action');
    }

    if (this.kioskMode) {
      const response = await this.kernel.browsers.playwright.execute(this.sessionId, {
        code: `await page.goto(${JSON.stringify(url)});`,
        timeout_sec: 60,
      });
      if (!response.success) {
        return { error: response.error ?? 'Playwright goto failed' };
      }
      await this.sleep(ACTION_DELAY_MS);
      return this.screenshot();
    }

    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: ['ctrl+l'],
    });
    await this.sleep(ACTION_DELAY_MS);

    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: ['ctrl+a'],
    });
    await this.sleep(100);

    await this.kernel.browsers.computer.typeText(this.sessionId, {
      text: url,
      delay: TYPING_DELAY_MS,
    });
    await this.sleep(ACTION_DELAY_MS);

    await this.kernel.browsers.computer.pressKey(this.sessionId, {
      keys: ['Return'],
    });

    await this.sleep(2000);
    return this.screenshot();
  }

  async screenshot(): Promise<ToolResult> {
    try {
      const response = await this.kernel.browsers.computer.captureScreenshot(this.sessionId);
      const blob = await response.blob();
      const arrayBuffer = await blob.arrayBuffer();
      const pngBuffer = Buffer.from(arrayBuffer);
      const webpBuffer = await sharp(pngBuffer).webp({ quality: 80 }).toBuffer();

      return {
        base64Image: webpBuffer.toString('base64'),
      };
    } catch (error) {
      throw new ToolError(`Failed to take screenshot: ${error}`);
    }
  }

  private getCoordinates(coords?: [number, number]): { x: number; y: number } {
    if (!coords || coords.length !== 2) {
      return { x: this.width / 2, y: this.height / 2 };
    }

    const [x, y] = coords;
    if (typeof x !== 'number' || typeof y !== 'number' || x < 0 || y < 0) {
      throw new ToolError(`Invalid coordinates: ${JSON.stringify(coords)}`);
    }

    return { x, y };
  }

  private mapKey(key: string): string {
    if (key.includes('+')) {
      const parts = key.split('+');
      const mappedParts = parts.map(part => {
        const trimmed = part.trim();
        const lower = trimmed.toLowerCase();
        
        if (MODIFIER_MAP[lower]) {
          return MODIFIER_MAP[lower];
        }
        
        return KEY_MAP[trimmed] || trimmed;
      });
      return mappedParts.join('+');
    }

    return KEY_MAP[key] || key;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
